feat: add Azure DevOps enterprise integration support

zerofltexx opened the pull request:
- [x] This change is worth documenting at https://docs.all-hands.dev/
- [x] Include this change in the Release Notes. If checked, you **must** provide an **end-user friendly** description for your change below

**End-user friendly description**

This PR adds enterprise/SaaS support for Azure DevOps integration. It builds on top of OpenHands/OpenHands/pull/11243 PR by adding SSO authentication, user identity mapping, and webhook security features required for multi-tenant SaaS deployment.

---
**Summary**

This PR extends the base Azure DevOps integration PR with enterprise-specific features.

**What This PR Adds**

1. **SSO Integration**
  - Keycloak identity provider configuration
  - Automatic background enrichment on user login
  
<img width="803" height="688" alt="Screenshot 2025-11-07 165250" src="https://github.com/user-attachments/assets/751059a5-44c0-408d-8269-bf6e55ca6b0f" />


2. **User Identity Resolution**
  - Resolves Azure AD Object ID → Azure DevOps User ID via Graph API
  - Stores mapping in Keycloak user attributes (`azure_devops_id`)
  - **Why needed:** Webhook payloads contain Azure DevOps User IDs (VSID), not Azure AD IDs. Without this mapping, we can't identify which OpenHands user triggered the webhook.

3. **Webhook Security**
  - Webhook verification using `(organization, subscriptionId, api_key)` lookup
  - Deduplication using Redis

4. **Enterprise Service Implementation**
  - Extends base `AzureDevOpsServiceImpl` with SaaS-specific features
  - Adds dual authentication support (OAuth JWT + PAT auto-detection)
  - Token manager integration for multi-user environments

5. **Webhook Event Processing**
  - Processes work item and PR comment webhooks
  - AI solvability analysis for work items
  - Acknowledgment messages posted back to Azure DevOps
  - User identity resolution from webhook payloads
  
<img width="1015" height="743" alt="Screenshot 2025-11-07 165156" src="https://github.com/user-attachments/assets/c1398549-38c5-4376-92cf-6b477a6e1262" />
<img width="1014" height="500" alt="Screenshot 2025-11-07 164824" src="https://github.com/user-attachments/assets/f28da4e6-da38-4d3d-8158-f039f8b7ea68" />


**Prerequisite**
Please follow the README.md in All-Hands-AI/OpenHands-Cloud/pull/238

Azure DevOps (Client side - unfortunately this can't be automated based on my research)
  - Add the application to ADO, this is easy if the app is within the same tenant for OH cloud self hosted case but for OH customers need to [create an enterprise application from a multitenant application](https://learn.microsoft.com/en-us/entra/identity/enterprise-apps/create-service-principal-cross-tenant?pivots=msgraph-powershell)
<img width="1263" height="927" alt="Screenshot 2025-11-07 165526" src="https://github.com/user-attachments/assets/5539715e-b88e-4325-9cbf-5ba3d237cec6" />

  - Create an [API key](https://docs.openhands.dev/openhands/usage/settings/api-keys-settings#create-api-key). Have this ready for the next step
 
  - Create the webhook, since it is at project level, this pair needs to be created in every project. In HTTP headers paste the API key like so:
<img width="1261" height="703" alt="Screenshot 2025-11-07 165842" src="https://github.com/user-attachments/assets/5cfe1be5-4b9b-4edb-8aa7-f1f19c92a5fa" />
<img width="1211" height="1223" alt="image" src="https://github.com/user-attachments/assets/c9a23d29-d6d7-4076-b8f1-81318f10c22f" />

**To-do**
- [X] Test webhook API key method
- [X] OpenHands-Cloud ADO PR
- [X] Test auto discover Organization from SSO

**Related Issues**

  - All-Hands-AI/OpenHands-Cloud#39

neubig reviewed (commented):
I'm happy to take a look at this when merge conflicts are resolved and tests are passing, please re-request review then!

neubig commented:
Hi, just a note, I'll circle back to reviewing this next week after we (hopefully) finish with shipping OpenHands v1 to the OpenHands cloud.

neubig commented:
We've shipped v1 so I'll be QAing this soon.

zerofltexx committed:
feat: add Azure DevOps enterprise integration support

This commit adds enterprise-grade Azure DevOps support with webhook-based automation
and service principal authentication for multi-tenant SaaS deployments.

Key changes:
- Implemented webhook event processing for work items and pull requests
- Added service principal authentication for automated operations
- Implemented Azure DevOps manager for handling webhook events
- Added work item and PR comment support with @mention detection
- Created callback processor for asynchronous webhook responses
- Added webhook registration and management endpoints
- Implemented Azure DevOps user enrichment and token management
- Added database schema for webhook storage
- Fixed API response format handling for comments and field values
- Removed linter suppressions for cleaner code

The implementation enables:
- Automated responses to work item assignments and @mentions
- Pull request comment handling with inline code review support
- Service principal-based API operations for scalability
- Multi-organization support with webhook deduplication
- Integration with Keycloak for user management
- Webhook-driven conversation creation and management

zerofltexx commented:
Updates:                                                                                                                                                                                                                 
  - Added V1 callback processor support for Azure DevOps (matching GitHub's V1 patterns)                                                                                                                                   
  - Added Azure DevOps button on the new login screen                                                                                                                                                                    
  - Added support for workitem.created webhook event type
  - Bumped migration number from 090 → 091 to resolve conflict with upstream main
  
Despite my best attempt, I can't make v1 work on my test environment, so I didn't test for said version. v0 works flawlessly still. Database already set to use v1, but there's a specific runtime-api & runtime pair needed? Appreciate some help on this then I can help test v1.
